lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs

```diff
@@ -558,6 +566,41 @@ impl FilesystemBackend {
         Ok(())
     }
 
+    /// Helper function for sending a [`tdp::SharedDirectoryReadRequest`] to the browser
+    /// and handling the [`tdp::SharedDirectoryReadResponse`] that is received in response.
+    fn tdp_sd_read(&mut self, rdp_req: efs::DeviceReadRequest) -> PduResult<()> {
+        if let Some(file) = self.file_cache.get(rdp_req.device_io_request.file_id) {
+            let tdp_req = tdp::SharedDirectoryReadRequest::from_fco(&rdp_req, file);
+            self.send_tdp_sd_read_request(tdp_req)?;
+            self.pending_sd_read_resp_handlers.insert(
+                rdp_req.device_io_request.completion_id,
+                SharedDirectoryReadResponseHandler::new(
+                    move |this: &mut FilesystemBackend,
+                          res: tdp::SharedDirectoryReadResponse|
+                          -> PduResult<()> {
+                        match res.err_code {
+                            TdpErrCode::Nil => this.send_read_response(
+                                rdp_req.device_io_request,
+                                NtStatus::SUCCESS,
+                                res.read_data,
+                            ),
+                            _ => this.send_read_response(
+                                rdp_req.device_io_request,
+                                NtStatus::UNSUCCESSFUL,
+                                vec![],
+                            ),
+                        }
+                    },
+                ),
```

Comment:
Can we extract this to a separate method to improve readability?

Reply:
8371f0153264cc6015da2011f18d72826306fc99